File: daedalus/redsim.py
"""Client for the ``redsim`` verifier.

The Rust core runs as a long-lived subprocess and speaks a length-prefixed
binary protocol over a pipe. That is a deliberate choice over a PyO3 extension:
a pipe needs no ABI to keep in step, no compiler toolchain at import time, and
no build step between cloning the repository and running the tests. Batching
amortises the syscall to nothing — §06 evaluates 64 candidates per spec, and a
batch of 64 is one round trip.

Typical use::

    with Verifier() as v:
        verdicts = v.evaluate_batch([grid_a, grid_b], placed_spec)

The client is not thread-safe; give each worker thread its own
:class:`Verifier`, or share one behind a lock.
"""

from __future__ import annotations

import os
import shutil
import struct
import subprocess
from dataclasses import dataclass
from pathlib import Path

from .grid import Grid
from .vocab import CELLS

MAGIC_REQ = b"RSIM"
MAGIC_RESP = b"RSOK"
PROTOCOL_VERSION = 2
OP_EVALUATE = 1
OP_POWER = 2

_REPO_ROOT = Path(__file__).resolve().parent.parent


class VerifierError(RuntimeError):
    """The worker could not be started, or spoke something unexpected."""


# --------------------------------------------------------------------------
# verdicts
# --------------------------------------------------------------------------


@dataclass(frozen=True, slots=True)
class RowMismatch:
    """One truth-table row that came out wrong."""

    inputs: int
    observed: int
    expected: int


#: Codes returned by the worker for :class:`Malformed`, in the order declared
#: by ``redsim::bin::malformed_code``.
MALFORMED_REASONS = (
    "floating_dust",
    "port_violation",
    "unsupported",
    "excluded_block",
    "masked_cell",
    "burnout",
    "history_dependent",
)

CONSTRAINT_NAMES = (None, "latency", "blocks", "region")


@dataclass(frozen=True, slots=True)
class Pass:
    latency_rt: int
    blocks: int
    bbox: tuple[int, int, int]

    kind = "pass"

    def is_pass(self) -> bool:
        return True

    def mismatch_count(self) -> int:
        return 0


@dataclass(frozen=True, slots=True)
class Fail:
    mismatched_rows: tuple[RowMismatch, ...]
    constraint: str | None
    #: What the layout measured, and the budget it missed. Scalar budgets use
    #: the first component only; a region uses both, as (x, z). Absent when no
    #: constraint was violated.
    got: tuple[int, int] | None = None
    budget: tuple[int, int] | None = None

    kind = "fail"

    def is_pass(self) -> bool:
        return False

    def overshoot(self) -> str | None:
        """The miss in words, for a caller reporting a near miss."""
        if self.constraint is None or self.got is None or self.budget is None:
            return None
        if self.constraint == "region":
            return (
                f"region {self.got[0]}x{self.got[1]} against a budget of "
                f"{self.budget[0]}x{self.budget[1]}"
            )
        unit = {"latency": "rt", "blocks": "blocks"}.get(self.constraint, "")
        measured = f"{self.got[0]} {unit}".strip()
        return f"{measured} against a budget of {self.budget[0]}"

    def mismatch_count(self) -> int:
        return len(self.mismatched_rows) + (1 if self.constraint else 0)


@dataclass(frozen=True, slots=True)
class Unstable:
    """Oscillating. Tracked apart from :class:`Fail` because a rising unstable
    rate during self-training is an early collapse warning, not the same
    signal as a rising failure rate."""

    period_ticks: int

    kind = "unstable"

    def is_pass(self) -> bool:
        return False

    def mismatch_count(self) -> int:
        return 1 << 30


@dataclass(frozen=True, slots=True)
class Malformed:
    reason: str
    at: tuple[int, int, int]

    kind = "malformed"

    def is_pass(self) -> bool:
        return False

    def mismatch_count(self) -> int:
        return 1 << 30


@dataclass(frozen=True, slots=True)
class PowerField:
    """The settled signal strengths for one input assignment.

    ``dust`` is one strength per cell, ``0..=15``, and zero anywhere that is
    not dust. The interesting reading is not the number itself but where it
    reaches zero: dust loses a step of strength per block, so a run that goes
    dark fifteen cells from its source needed a repeater.
    """

    dust: list[int]
    settled: bool
    game_ticks: int
    #: Which output lamps came on, as a bitmask over the declared outputs.
    outputs: int

    def reach(self) -> int:
        """How many cells are carrying any signal at all."""
        return sum(1 for level in self.dust if level)


Verdict = Pass | Fail | Unstable | Malformed


# --------------------------------------------------------------------------
# locating the binary
# --------------------------------------------------------------------------


def _target_binaries(target: Path, os_name: str = os.name) -> tuple[Path, Path]:
    """Return Cargo's release and debug verifier paths for this platform.

    Cargo appends ``.exe`` to Windows binaries.  Keeping the platform detail
    here makes the discovery rule testable on every CI host instead of only on
    a Windows runner.
    """
    filename = "redsim.exe" if os_name == "nt" else "redsim"
    return target / "release" / filename, target / "debug" / filename


def find_binary(explicit: str | os.PathLike[str] | None = None) -> Path:
    """Locate the ``redsim`` executable.

    Search order: an explicit path, ``$REDSIM_BIN``, the cargo release and
    debug target directories, then ``$PATH``. Release is preferred because the
    debug build is roughly twenty times slower and would quietly make every
    throughput number in the README wrong.
    """
    candidates: list[Path] = []
    if explicit:
        candidates.append(Path(explicit))
    env = os.environ.get("REDSIM_BIN")
    if env:
        candidates.append(Path(env))
    target = Path(os.environ.get("CARGO_TARGET_DIR", _REPO_ROOT / "target"))
    candidates += _target_binaries(target)
    found = shutil.which("redsim")
    if found:
        candidates.append(Path(found))

    for c in candidates:
        if c.is_file() and os.access(c, os.X_OK):
            return c
    raise VerifierError(
        "could not find the redsim binary. Build it with "
        "`cargo build --release -p redsim`, or set $REDSIM_BIN."
    )


# --------------------------------------------------------------------------
# the client
# --------------------------------------------------------------------------


class Verifier:
    """A handle on a running ``redsim`` worker."""

    def __init__(self, binary: str | os.PathLike[str] | None = None):
        self.binary = find_binary(binary)
        self._proc: subprocess.Popen[bytes] | None = None
        #: Total grids evaluated, so callers can report verifier throughput as
        #: the scaling axis it actually is.
        self.evaluated = 0

    # -- lifecycle ---------------------------------------------------------

    def start(self) -> None:
        if self._proc is not None and self._proc.poll() is None:
            return
        try:
            self._proc = subprocess.Popen(
                [str(self.binary), "serve"],
                stdin=subprocess.PIPE,
                stdout=subprocess.PIPE,
                stderr=None,
                bufsize=0,
            )
        except OSError as e:  # pragma: no cover - environment specific
            raise VerifierError(f"could not start {self.binary}: {e}") from e

    def close(self) -> None:
        """Shut the worker down and close both pipes.

        Closing stdin alone leaks the stdout descriptor. One leak is harmless;
        a training round that opens a verifier per worker thread runs out of
        descriptors, and it does so a long way from the cause.
        """
        if self._proc is None:
            return
        proc, self._proc = self._proc, None
        try:
            if proc.stdin:
                proc.stdin.close()
            proc.wait(timeout=10)
        except Exception:  # pragma: no cover - best effort teardown
            proc.kill()
            proc.wait(timeout=5)
        finally:
            for pipe in (proc.stdout, proc.stderr):
                if pipe is not None and not pipe.closed:
                    pipe.close()

    def __enter__(self) -> Verifier:
        self.start()
        return self

    def __exit__(self, *exc) -> None:
        self.close()

    # -- protocol ----------------------------------------------------------

    def _write(self, data: bytes) -> None:
        """Write all of ``data``, however many syscalls that takes.

        The worker is started unbuffered, so this pipe is a raw file object
        and a single write is free to consume only part of the buffer. A
        request is comfortably larger than a pipe (a batch of 64 grids is
        ~98 KB against a 64 KB pipe), so the short write is the normal case
        rather than the rare one.
        """
        assert self._proc is not None and self._proc.stdin is not None
        view = memoryview(data)
        while view:
            written = self._proc.stdin.write(view)
            if not written:
                raise VerifierError("redsim worker stopped accepting input")
            view = view[written:]

    def _read_exact(self, n: int) -> bytes:
        """Read exactly ``n`` bytes, or explain why it could not.

        Same reason as the write side: an unbuffered read returns whatever
        has arrived, which is not necessarily what was asked for. Treating a
        short read as a dead worker misreports a perfectly healthy one, and
        does it intermittently, which is the worst way to find out.
        """
        assert self._proc is not None and self._proc.stdout is not None
        buf = bytearray()
        while len(buf) < n:
            chunk = self._proc.stdout.read(n - len(buf))
            if not chunk:
                code = self._proc.poll()
                message = f"redsim worker closed the pipe after {len(buf)}/{n} bytes"
                if code is not None:
                    message += f" (exit {code})"
                if code not in (None, 0) and not buf:
                    # Dying before writing a single byte is what a worker does
                    # when it rejects the request outright, and by far the
                    # commonest reason is a binary older than this client.
                    # The worker says so on stderr, which is easy to miss and
                    # is swallowed entirely by a test runner or a web server.
                    message += (
                        f"\nThe worker refused the request before answering. This "
                        f"client speaks protocol {PROTOCOL_VERSION}; an older "
                        f"binary will not.\nRebuild it: cargo build --release -p redsim"
                    )
                raise VerifierError(message)
            buf += chunk
        return bytes(buf)

    def evaluate_batch(self, grids, spec) -> list[Verdict]:
        """Evaluate many candidate grids against one placed spec.

        ``grids`` may be :class:`~daedalus.grid.Grid` objects, ``bytes`` or any
        iterable of ``CELLS`` token ids.
        """
        payloads = [_as_payload(g) for g in grids]
        if not payloads:
            return []
        self.start()

        req = bytearray(MAGIC_REQ)
        req += bytes([PROTOCOL_VERSION, OP_EVALUATE])
        req += _encode_spec(spec)
        req += struct.pack("<I", len(payloads))
        for p in payloads:
            req += p
        self._write(bytes(req))

        magic = self._read_exact(4)
        if magic != MAGIC_RESP:
            raise VerifierError(f"bad response magic {magic!r}")
        (n,) = struct.unpack("<I", self._read_exact(4))
        if n != len(payloads):
            raise VerifierError(f"asked for {len(payloads)} verdicts, got {n}")
        out = [self._read_verdict() for _ in range(n)]
        self.evaluated += n
        return out

    def evaluate(self, grid, spec) -> Verdict:
        return self.evaluate_batch([grid], spec)[0]

    def power(self, grid, spec, assignment: int = 0) -> PowerField:
        """Settle the circuit with one input assignment and read the levels.

        A verdict says a circuit is wrong. This says *where the signal
        stopped*, which is the question anyone looking at a broken layout is
        actually asking. The simulator computes these levels on the way to
        every verdict already; this stops discarding them.

        ``assignment`` is a bitmask over the declared inputs, in order.
        """
        self.start()
        req = bytearray(MAGIC_REQ)
        req += bytes([PROTOCOL_VERSION, OP_POWER])
        req += _encode_spec(spec)
        req += struct.pack("<Q", assignment)
        req += _as_payload(grid)
        self._write(bytes(req))

        magic = self._read_exact(4)
        if magic != MAGIC_RESP:
            raise VerifierError(f"bad response magic {magic!r}")
        status = self._read_exact(1)[0]
        if status:
            raise VerifierError("this grid is too malformed to simulate")
        settled = bool(self._read_exact(1)[0])
        (ticks,) = struct.unpack("<I", self._read_exact(4))
        (outputs,) = struct.unpack("<Q", self._read_exact(8))
        dust = list(self._read_exact(CELLS))
        return PowerField(dust=dust, settled=settled, game_ticks=ticks, outputs=outputs)

    def _read_verdict(self) -> Verdict:
        kind = self._read_exact(1)[0]
        if kind == 0:
            latency, blocks, bx, by, bz = struct.unpack("<BHBBB", self._read_exact(6))
            return Pass(latency_rt=latency, blocks=blocks, bbox=(bx, by, bz))
        if kind == 1:
            (count,) = struct.unpack("<H", self._read_exact(2))
            rows = []
            for _ in range(count):
                inp, obs, exp = struct.unpack("<QQQ", self._read_exact(24))
                rows.append(RowMismatch(inp, obs, exp))
            code = self._read_exact(1)[0]
            if code >= len(CONSTRAINT_NAMES):
                raise VerifierError(f"unknown constraint code {code}")
            if code == 0:
                return Fail(tuple(rows), None)
            got_a, got_b, max_a, max_b = struct.unpack("<4I", self._read_exact(16))
            return Fail(tuple(rows), CONSTRAINT_NAMES[code], (got_a, got_b), (max_a, max_b))
        if kind == 2:
            return Unstable(period_ticks=self._read_exact(1)[0])
        if kind == 3:
            code, x, y, z = struct.unpack("<BBBB", self._read_exact(4))
            if code >= len(MALFORMED_REASONS):
                raise VerifierError(f"unknown malformed code {code}")
            return Malformed(MALFORMED_REASONS[code], (x, y, z))
        raise VerifierError(f"unknown verdict kind {kind}")


def _as_payload(g) -> bytes:
    if isinstance(g, Grid):
        return g.to_bytes()
    if isinstance(g, (bytes, bytearray, memoryview)):
        b = bytes(g)
    else:
        b = bytes(bytearray(g))
    if len(b) != CELLS:
        raise ValueError(f"grid needs exactly {CELLS} tokens, got {len(b)}")
    return b


def _encode_spec(spec) -> bytes:
    """Serialise a :class:`daedalus.spec.PlacedSpec` for the worker."""
    out = bytearray()
    out.append(len(spec.input_ports))
    for x, y, z in spec.input_ports:
        out += bytes((x, y, z))
    out.append(len(spec.output_ports))
    for x, y, z in spec.output_ports:
        out += bytes((x, y, z))
    rows = spec.rows
    out += struct.pack("<I", len(rows))
    for r in rows:
        out += struct.pack("<Q", r)

    c = spec.constraints
    flags = 0
    if c.max_latency_rt is not None:
        flags |= 1
    if c.max_blocks is not None:
        flags |= 2
    if c.max_region is not None:
        flags |= 4
    region = c.max_region or (0, 0)
    out.append(flags)
    out += struct.pack("<I", c.max_latency_rt or 0)
    out += struct.pack("<H", c.max_blocks or 0)
    out += bytes((region[0], region[1]))
    return bytes(out)


#: A module-level worker for scripts and notebooks that just want to call
#: `evaluate` without managing a context manager.
_shared: Verifier | None = None


def shared() -> Verifier:
    global _shared
    if _shared is None:
        _shared = Verifier()
        _shared.start()
    return _shared


def evaluate(grid, spec) -> Verdict:
    return shared().evaluate(grid, spec)


def evaluate_batch(grids, spec) -> list[Verdict]:
    return shared().evaluate_batch(grids, spec)

File: crates/redsim/src/bin/redsim.rs
//! `redsim` — the verifier as a long-lived worker process.
//!
//! Python drives the simulator over a length-prefixed binary protocol on
//! stdin/stdout rather than through a native extension. That is a deliberate
//! trade: PyO3 would shave a few microseconds per call, but it also means a
//! compiler toolchain, an ABI to keep in step, and a build step between
//! "clone the repo" and "run the tests". A pipe has none of that, batches
//! amortise the syscall away, and the same binary is usable from any language.
//!
//! Protocol, all integers little-endian:
//!
//! ```text
//! request  := "RSIM" u8:version u8:op spec u32:n_grids (1536 bytes)*
//! spec     := u8:n_in  (u8 x, u8 y, u8 z)*
//!             u8:n_out (u8 x, u8 y, u8 z)*
//!             u32:n_rows u64:row*
//!             u8:flags u32:max_latency u16:max_blocks u8:region_x u8:region_z
//! response := "RSOK" u32:n_verdicts verdict*
//! verdict  := u8:kind ...                     -- see `write_verdict`
//! ```

use std::io::{self, BufWriter, Read, Write};

use redsim::grid::CELLS;
use redsim::spec::{Constraints, Port};
use redsim::verdict::{ConstraintViolation, MalformedReason};
use redsim::{evaluate_batch, Pos, Spec, Verdict};

const MAGIC_REQ: &[u8; 4] = b"RSIM";
const MAGIC_RESP: &[u8; 4] = b"RSOK";
const PROTOCOL_VERSION: u8 = 2;
const OP_EVALUATE: u8 = 1;

fn main() {
    let args: Vec<String> = std::env::args().skip(1).collect();
    match args.first().map(String::as_str) {
        None | Some("serve") => {
            if let Err(e) = serve() {
                if e.kind() != io::ErrorKind::UnexpectedEof {
                    eprintln!("redsim: {e}");
                    std::process::exit(1);
                }
            }
        }
        Some("--version") | Some("version") => {
            println!("redsim {}", env!("CARGO_PKG_VERSION"));
        }
        Some("selftest") => selftest(),
        Some("vocab") => dump_vocab(),
        Some(other) => {
            eprintln!("redsim: unknown command {other:?}");
            eprintln!("usage: redsim [serve|selftest|vocab|version]");
            std::process::exit(2);
        }
    }
}

/// Read requests until stdin closes. One process serves an entire training
/// round; the cost of spawning is paid once, not per candidate.
fn serve() -> io::Result<()> {
    let stdin = io::stdin();
    let mut r = stdin.lock();
    let stdout = io::stdout();
    let mut w = BufWriter::new(stdout.lock());

    loop {
        let mut magic = [0u8; 4];
        if !read_exact_or_eof(&mut r, &mut magic)? {
            return Ok(());
        }
        if &magic != MAGIC_REQ {
            return Err(io::Error::new(io::ErrorKind::InvalidData, "bad request magic"));
        }
        let version = read_u8(&mut r)?;
        if version != PROTOCOL_VERSION {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                format!("protocol version {version} != {PROTOCOL_VERSION}"),
            ));
        }
        let op = read_u8(&mut r)?;
        if op != OP_EVALUATE {
            return Err(io::Error::new(io::ErrorKind::InvalidData, format!("unknown op {op}")));
        }

        let spec = read_spec(&mut r)?;
        let n = read_u32(&mut r)? as usize;
        let mut grids = Vec::with_capacity(n);
        for _ in 0..n {
            let mut buf = vec![0u8; CELLS];
            r.read_exact(&mut buf)?;
            grids.push(buf);
        }

        let verdicts = evaluate_batch(&grids, &spec);

        w.write_all(MAGIC_RESP)?;
        w.write_all(&(verdicts.len() as u32).to_le_bytes())?;
        for v in &verdicts {
            write_verdict(&mut w, v)?;
        }
        w.flush()?;
    }
}

fn read_exact_or_eof<R: Read>(r: &mut R, buf: &mut [u8]) -> io::Result<bool> {
    let mut filled = 0;
    while filled < buf.len() {
        match r.read(&mut buf[filled..])? {
            0 if filled == 0 => return Ok(false),
            0 => return Err(io::ErrorKind::UnexpectedEof.into()),
            n => filled += n,
        }
    }
    Ok(true)
}

fn read_u8<R: Read>(r: &mut R) -> io::Result<u8> {
    let mut b = [0u8; 1];
    r.read_exact(&mut b)?;
    Ok(b[0])
}

fn read_u16<R: Read>(r: &mut R) -> io::Result<u16> {
    let mut b = [0u8; 2];
    r.read_exact(&mut b)?;
    Ok(u16::from_le_bytes(b))
}

fn read_u32<R: Read>(r: &mut R) -> io::Result<u32> {
    let mut b = [0u8; 4];
    r.read_exact(&mut b)?;
    Ok(u32::from_le_bytes(b))
}

fn read_u64<R: Read>(r: &mut R) -> io::Result<u64> {
    let mut b = [0u8; 8];
    r.read_exact(&mut b)?;
    Ok(u64::from_le_bytes(b))
}

fn read_ports<R: Read>(r: &mut R, prefix: &str) -> io::Result<Vec<Port>> {
    let n = read_u8(r)? as usize;
    let mut out = Vec::with_capacity(n);
    for k in 0..n {
        let x = read_u8(r)? as i32;
        let y = read_u8(r)? as i32;
        let z = read_u8(r)? as i32;
        out.push(Port { name: format!("{prefix}{k}"), pos: Pos::new(x, y, z) });
    }
    Ok(out)
}

fn read_spec<R: Read>(r: &mut R) -> io::Result<Spec> {
    let inputs = read_ports(r, "in")?;
    let outputs = read_ports(r, "out")?;
    let n_rows = read_u32(r)? as usize;
    let mut rows = Vec::with_capacity(n_rows);
    for _ in 0..n_rows {
        rows.push(read_u64(r)?);
    }
    let flags = read_u8(r)?;
    let latency = read_u32(r)?;
    let blocks = read_u16(r)?;
    let rx = read_u8(r)?;
    let rz = read_u8(r)?;
    let constraints = Constraints {
        max_latency_rt: (flags & 1 != 0).then_some(latency),
        max_blocks: (flags & 2 != 0).then_some(blocks),
        max_region: (flags & 4 != 0).then_some((rx, rz)),
    };
    Spec::new(inputs, outputs, rows, constraints)
        .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e.to_string()))
}

fn malformed_code(reason: &MalformedReason) -> (u8, Pos) {
    match reason {
        MalformedReason::FloatingDust { at } => (0, *at),
        MalformedReason::PortViolation { at } => (1, *at),
        MalformedReason::Unsupported { at } => (2, *at),
        MalformedReason::ExcludedBlock { at } => (3, *at),
        MalformedReason::MaskedCell { at } => (4, *at),
        MalformedReason::Burnout { at } => (5, *at),
        MalformedReason::HistoryDependent => (6, Pos::new(0, 0, 0)),
    }
}

fn write_verdict<W: Write>(w: &mut W, v: &Verdict) -> io::Result<()> {
    match v {
        Verdict::Pass { latency_rt, blocks, bbox } => {
            w.write_all(&[0, *latency_rt])?;
            w.write_all(&blocks.to_le_bytes())?;
            w.write_all(&[bbox.0, bbox.1, bbox.2])?;
        }
        Verdict::Fail { mismatched_rows, constraint } => {
            w.write_all(&[1])?;
            w.write_all(&(mismatched_rows.len() as u16).to_le_bytes())?;
            for m in mismatched_rows {
                w.write_all(&m.inputs.to_le_bytes())?;
                w.write_all(&m.observed.to_le_bytes())?;
                w.write_all(&m.expected.to_le_bytes())?;
            }
            // The code alone says which budget was missed but not by how
            // much, and the numbers are right here. "38 blocks against a
            // budget of 34" is something a caller can act on; "blocks" is not.
            // Scalar budgets leave the second component zero; a region uses
            // both, as (x, z).
            let (code, got, max) = match constraint {
                None => (0u8, (0u32, 0u32), (0u32, 0u32)),
                Some(ConstraintViolation::Latency { got, max }) => {
                    (1u8, (*got, 0), (*max, 0))
                }
                Some(ConstraintViolation::Blocks { got, max }) => {
                    (2u8, (u32::from(*got), 0), (u32::from(*max), 0))
                }
                Some(ConstraintViolation::Region { got, max }) => (
                    3u8,
                    (u32::from(got.0), u32::from(got.1)),
                    (u32::from(max.0), u32::from(max.1)),
                ),
            };
            w.write_all(&[code])?;
            if code != 0 {
                for v in [got.0, got.1, max.0, max.1] {
                    w.write_all(&v.to_le_bytes())?;
                }
            }
        }
        Verdict::Unstable { period_ticks } => w.write_all(&[2, *period_ticks])?,
        Verdict::Malformed { reason } => {
            let (code, at) = malformed_code(reason);
            w.write_all(&[3, code, at.x as u8, at.y as u8, at.z as u8])?;
        }
    }
    Ok(())
}

/// Dump the vocabulary and a set of reference spec hashes as text.
///
/// The Python side re-derives all of this independently, and
/// `tests/test_vocab_parity.py` diffs the two. Token ids are a wire format
/// baked into every serialised corpus and every checkpoint, so "the two
/// implementations agree" has to be a test, not a comment.
fn dump_vocab() {
    use redsim::block::{Block, CONTROL_BASE, VOCAB_SIZE};
    use redsim::grid::glyph;

    println!("# id\tglyph\topaque\tconductive\tstate");
    for t in 0..CONTROL_BASE {
        let b = Block::from_token(t).expect("non-control ids decode");
        println!(
            "{t}\t{}\t{}\t{}\t{}",
            glyph(b),
            u8::from(b.is_opaque()),
            u8::from(b.is_conductive()),
            b.state_string()
        );
    }
    for t in CONTROL_BASE..VOCAB_SIZE as u8 {
        let name = match t {
            redsim::block::TOK_PAD => "PAD",
            redsim::block::TOK_BOS => "BOS",
            redsim::block::TOK_EOS => "EOS",
            _ => "MASK",
        };
        println!("{t}\t-\t0\t0\tcontrol:{name}");
    }

    println!("# geometry\tsx\tsy\tsz\tcells\tsubstrate_y\tlogic_y\tinput_x\toutput_x");
    println!(
        "geometry\t{}\t{}\t{}\t{}\t{}\t{}\t{}\t{}",
        redsim::grid::SX,
        redsim::grid::SY,
        redsim::grid::SZ,
        redsim::grid::CELLS,
        redsim::grid::SUBSTRATE_Y,
        redsim::grid::LOGIC_Y,
        redsim::grid::INPUT_X,
        redsim::grid::OUTPUT_X
    );

    // Reference semantic hashes. Any drift here means the corpus dedup key
    // has silently changed and every cached dataset is stale.
    println!("# hash\tn_in\tn_out\trows...\tvalue");
    for (n_in, n_out, rows) in [
        (1usize, 1usize, vec![0u64, 1]),
        (1, 1, vec![1, 0]),
        (2, 1, vec![1, 1, 1, 0]),
        (2, 1, vec![0, 0, 0, 1]),
        (2, 1, vec![0, 1, 1, 0]),
        (2, 2, vec![0b00, 0b01, 0b10, 0b11]),
        (3, 1, vec![0, 1, 1, 0, 1, 0, 0, 1]),
    ] {
        let inputs: Vec<Port> = (0..n_in)
            .map(|k| Port { name: format!("in{k}"), pos: Pos::new(0, 1, 2 * k as i32 + 1) })
            .collect();
        let outputs: Vec<Port> = (0..n_out)
            .map(|k| Port { name: format!("out{k}"), pos: Pos::new(15, 1, 2 * k as i32 + 1) })
            .collect();
        let spec = Spec::new(inputs, outputs, rows.clone(), Constraints::default()).unwrap();
        let cells: Vec<String> = rows.iter().map(|r| r.to_string()).collect();
        println!("hash\t{n_in}\t{n_out}\t{}\t{:016x}", cells.join(","), spec.semantic_hash());
    }
}

/// A five-second sanity check that the binary works at all, for anyone who
/// just cloned the repo and wants to know before reading further.
fn selftest() {
    use redsim::builder::Builder;
    use redsim::{evaluate, Constraints};

    // NAND = !(A & B) = !A | !B. Invert each input with a torch, then merge
    // the two dust runs — a dust join is a free OR.
    let mut b = Builder::new();
    b.input("A", 6);
    b.dust_x(2, 3, 6);
    b.invert(4, 6);
    b.dust_x(6, 7, 6);

    b.input("B", 10);
    b.dust_x(2, 3, 10);
    b.invert(4, 10);
    b.dust_x(6, 7, 10);

    b.dust_z(7, 6, 10);
    b.dust_x(8, 13, 8);
    b.output("Q", 8);

    let spec = b.spec(vec![1, 1, 1, 0], Constraints::default());
    let verdict = evaluate(&b.grid, &spec);
    println!("{:?}", b.grid);
    println!("NAND -> {verdict}");
    if !verdict.is_pass() {
        eprintln!("selftest FAILED");
        std::process::exit(1);
    }
    println!("selftest ok");
}
